src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs

```diff
@@ -0,0 +1,59 @@
+using BlazorWasm_CSharp.Shared.Models;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Authorization;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace BlazorWasm_CSharp.Client
+{
+    public class IdentityAuthenticationStateProvider : AuthenticationStateProvider
+    {
+        private ClaimsPrincipal _currentUser;
+
+        public IdentityAuthenticationStateProvider(HttpClient client)
+        {
+            Client = client;
+        }
+
+        public HttpClient Client { get; }
+
+        public override async Task<AuthenticationState> GetAuthenticationStateAsync()
+        {
+            if (_currentUser != null)
+            {
+                return new AuthenticationState(_currentUser);
+            }
+
+            await UpdateUser();
+
+            return new AuthenticationState(_currentUser);
```

Comment:
Not sure if there's a reason for the way this is structured. Is the following simplification acceptable?

```cs
        public override async Task<AuthenticationState> GetAuthenticationStateAsync()
        {
            if (_currentUser == null)
            {
                await UpdateUser();
            }

            return new AuthenticationState(_currentUser);
}
```

Also I'd have a minor preference for replacing `UpdateUser` with `FetchUser` and having it return a `ClaimsPrincipal`, so then we have:

```cs
_currentUser = await FetchUser();
```

... so that all the concerns about `_currentUser` are scoped to a single method. Not a super big deal though!

Reply:
This was updated, not sure why the diff doesn't show it.

Reply:
Updated, not sure why this doesn't reflect it.